plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesDialog.java

```diff
@@ -0,0 +1,116 @@
+import org.eclipse.ui.forms.widgets.ExpandableComposite;
+
+import com.google.cloud.tools.eclipse.appengine.deploy.ui.DeployPreferencesPanel.SectionExpansionHandler;
+import com.google.cloud.tools.eclipse.appengine.ui.AppEngineImages;
+import com.google.common.base.Preconditions;
+
+public class DeployPreferencesDialog extends TitleAreaDialog {
+
+  // if the image is smaller (e.g. 32x32, it will break the layout of the TitleAreaDialog)
+  // seems like an Eclipse/JFace bug
+  private Image titleImage = AppEngineImages.appEngine(64).createImage();
+
+  private DeployPreferencesPanel content;
+  private IProject project;
+
+  public DeployPreferencesDialog(Shell parentShell, IProject project) {
+    super(parentShell);
+
+    Preconditions.checkNotNull(project, "project is null");
+    this.project = project;
+  }
+
+  @Override
+  protected Control createContents(Composite parent) {
+    Control contents = super.createContents(parent);
+
+    getShell().setText(Messages.getString("deploy.preferences.dialog.title"));
+    setTitle(Messages.getString("deploy.preferences.dialog.title.withProject", project.getName()));
+
+    if (titleImage != null) {
+      setTitleImage(titleImage);
+    }
+
+    // TitleAreaDialogSupport does not validate initially, let's trigger validation this way
+    content.getDataBindingContext().updateTargets();
+
+    return contents;
+  }
+  
+  @Override
+  protected Control createDialogArea(final Composite parent) {
+    Composite dialogArea = (Composite) super.createDialogArea(parent);
+    setMargin(dialogArea);
+    content = new DeployPreferencesPanel(dialogArea, project, getExpansionHandler());
+    GridDataFactory.fillDefaults().grab(true, false).applyTo(content);
+    TitleAreaDialogSupport.create(this, content.getDataBindingContext()).setValidationMessageProvider(new ValidationMessageProvider() {
+      @Override
+      public int getMessageType(ValidationStatusProvider statusProvider) {
+        int type = super.getMessageType(statusProvider);
+        Button okButton = getButton(IDialogConstants.OK_ID);
+        if (okButton != null) {
+          okButton.setEnabled(type != IMessageProvider.ERROR);
+        }
+        return type;
+      }
+    });
+    return dialogArea;
+  }
+
+  private void setMargin(Composite dialogArea) {
+    Layout layout = dialogArea.getLayout();
+    if (layout instanceof GridLayout) {
+      ((GridLayout)layout).marginWidth = 5;
+    }
+  }
+  
+  private SectionExpansionHandler getExpansionHandler() {
+    return new SectionExpansionHandler() {
+
+      @Override
+      public void handleExpansionEvent(ExpansionEvent event) {
+        Shell shell = getShell();
+        shell.setMinimumSize(shell.getSize().x, 0);
+        shell.pack();
+        ((ExpandableComposite) event.getSource()).getParent().layout();
+        shell.setMinimumSize(shell.getSize());
+      }
+    };
+  }
+
+  @Override
+  protected void okPressed() {
+    content.savePreferences();
```

Comment:
An aside, it feels a bit strange that a the *panel* knows how to persist the model. It would seem the responsibility of the dialog or the dialog's creator.